executor/executor.go

```diff
@@ -122,22 +143,92 @@ func (e *Executor) Run(conflicts state.Keys, f func() error) {
-		if ok {
-			lt := e.tasks[latest]
-			if !lt.executed {
-				if t.dependencies == nil {
-					t.dependencies = set.NewSet[int](defaultSetSize)
+	for k, v := range conflicts {
+		// Get last blocked transaction
+		parent, exists := e.edges[k]
+		if !exists {
+			concurrentReads := set.NewSet[int](defaultSetSize)
+			if v == state.Read {
+				// Record any reads on the same key to help
+				// with updating dependencies properly
+				concurrentReads.Add(id)
+			}
+			// Key doesn't exist, so we add it to our edge map
+			e.edges[k] = &KeyData{Parent: id, Permissions: v, ConcurrentReads: concurrentReads}
+			continue
+		}
+		pt := e.tasks[parent.Parent]
+		if !pt.executed {
+			if t.dependencies == nil {
+				t.dependencies = set.NewSet[int](defaultSetSize)
+			}
+			if pt.blocking == nil {
+				pt.blocking = set.NewSet[int](defaultSetSize)
+			}
+			// key has ONLY a Read permission
+			if v == state.Read {
+				parent.ConcurrentReads.Add(id)
+				// If the first read hasn't executed for some reason, and
+				// we're doing a bunch of Reads with no conflicts like
+				// R->R->R->..., we don't want to record any dependencies
+				if parent.Permissions == state.Read {
+					continue
 				}
-				t.dependencies.Add(lt.id)
-				if lt.blocking == nil {
-					lt.blocking = set.NewSet[int](defaultSetSize)
+				// Last blocked transaction was a Allocate/Write
+				pt.blocking.Add(id)
+				t.dependencies.Add(parent.Parent)
+				continue
+			}
+
+			// key contains an Allocate/Write permission
+			if v.Has(state.Allocate) || v.Has(state.Write) {
+				// This may occur if we're doing a lot of W->W->W->...,
+				// we still want to record that we're blocked
+				if parent.ConcurrentReads.Len() == 0 {
+					pt.blocking.Add(id)
+					t.dependencies.Add(parent.Parent)
+				} else {
+					// With a bunch of reads before our write,
+					// we need to update that we're blocked by
+					// all of these reads
+					for b := range parent.ConcurrentReads {
+						bt := e.tasks[b]
+						// In the case one of the concurrent reads have
+						// executed and we're at this stage of the code,
+						// we only want to record that we're blocked on
+						// the concurrent reads that are not yet executed.
+						if !bt.executed {
+							bt.blocking.Add(id)
+							t.dependencies.Add(b)
+						}
+					}
+					// Any subsequent reads will now be blocked by this
+					// task, if it hasn't executed yet.
+					parent.ConcurrentReads.Clear()
+				}
+			}
+		} else {
+			// We don't need to record the case of W->W->W->...,
+			// since that parent write already executed. We
+			// consider any outstanding reads that still need
+			// to be executed once its parent write ran.
+			if v.Has(state.Allocate) || v.Has(state.Write) {
+				for b := range parent.ConcurrentReads {
+					bt := e.tasks[b]
+					if !bt.executed {
+						bt.blocking.Add(id)
+						t.dependencies.Add(b)
+					}
 				}
-				lt.blocking.Add(id)
+			}
+			parent.ConcurrentReads.Clear()
```

Comment:
Why would you want to clear `ConcurrentReads` here? Seems like it should be included in the `if` above?

Reply:
reposting for visibility:

we want to clear concurrentReads here before updating because if we had the case of only reads like: R->R->R->R, and the first R already executed, concurrentReads was populated with that first R, but was already done by the time we reached this part of the code, so we need to reset it to avoid adding unwanted concurrentReads